aspnetcore/signalr/hub-filters.md

```diff
@@ -0,0 +1,203 @@
+
+```csharp
+public class CustomFilter : IHubFilter
+{
+    public async ValueTask<object> InvokeMethodAsync(
+        HubInvocationContext invocationContext, Func<HubInvocationContext, ValueTask<object>> next)
+    {
+        Console.WriteLine($"Calling hub method '{invocationContext.HubMethodName}'");
+        try
+        {
+            return await next(invocationContext);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine($"Exception calling '{invocationContext.HubMethodName}'");
+            throw ex;
+        }
+    }
+
+    // Optional method
+    public Task OnConnectedAsync(HubLifetimeContext context, Func<HubLifetimeContext, Task> next)
+    {
+        return next(context);
+    }
+
+    // Optional method
+    public Task OnDisconnectedAsync(
+        HubLifetimeContext context, Exception exception, Func<HubLifetimeContext, Exception, Task> next)
+    {
+        return next(context, exception);
+    }
+}
+```
+
+Filters are very similar to middleware. The `next` method invokes the next filter. The final filter will invoke the hub method. Filters can also store the result from awaiting `next` and run logic after the hub method has been called before returning the result from `next`.
+
+To skip a hub method invocation in a filter, throw an exception of type `HubException` instead of calling `next`. The client will receive an error if it was expecting a result.
+
+## Use hub filters
+
+When writing the filter logic, try to make it generic by using attributes on hub methods instead of checking for hub method names.
+
+Consider a filter that will check a hub method argument for banned phrases and replace any phrases it finds with `***`.
+For this example, assume a `LanguageFilterAttribute` class is defined. The class has a property named `FilterArgument` that can be set when using the attribute.
+
+1. Place the attribute on the hub method that has a string argument to be cleaned:
+
+    ```csharp
+    public class ChatHub
+    {
+        [LanguageFilter(filterArgument: 0)]
+        public async Task SendMessage(string message, string username)
+        {
+            await Clients.All.SendAsync("SendMessage", $"{username} says: {message}");
+        }
+    }
+    ```
+
+1. Define a hub filter to check for the attribute and replace banned phrases in a hub method argument with `***`:
+
+    ```csharp
+    public class LanguageFilter : IHubFilter
+    {
+        // populated from a file or inline
+        private List<string> bannedPhrases = new List<string> { "async void", ".Result" };
+
+        public async ValueTask<object> InvokeMethodAsync(HubInvocationContext invocationContext, Func<HubInvocationContext, ValueTask<object>> next)
```

Comment:
Move the method parameters to the next line to prevent horizontal scrolling in the sample.